### ISSUE TITLE
Export Stream class so that we can extend it with our own thingies

### ISSUE THREAD

**Issue by balupton** [caolan/highland#70]
Right now we are limited to only using what highland provides, it would be great if we could extend highland somehow. Either by highland exposing the stream class so we can add our own things to it, or by providing a utility method like `_.add('uniq', function(){})`

**Comment by loveencounterflow** [comment]
suggestion:

```js
_([1,2,3]).use(foo).use(bar).frobulate().brobdignize()
```
where `foo` and `bar` are plain objects whose function members will be used as mix-ins all the way down the stream. not sure whether it is technically feasible with the current implementation, but it does have the advantage of circumventing, on the user side, those very hairy prototype issues that JS is fraught with. instead, just set up a toolbox as `var foo = { frobulate: function ( ... ) { ... } }` and you're good to go.

the `use` keyword has become very frequent usage amongst 'language' programmers to indicate what features to use in cases of language extension / modification; one example is JS `"use strict"`.

one could also start a module with

```js
var _ = require( 'highland' );
_ = _.use( require( 'foo' ) );
```
to obtain a bespoke highland instance. 

lastly, in order not to make this an afterthought and to capture the spirit of the Atom editor's setup (where 'almost everything' the editor can do is implemented as an extension), it should be possible to make it so that even the highland standard edition is composed, say, of `_ = _.use(core).use(stdlib)`.

**Comment by loveencounterflow** [comment]
just realized there's a closed pull request in issue #71 that provides a candidate implementation. 

on the plus side, it's like the simplest conceivable solution:

```js
_.plugin = function(key, value){
	Stream.prototype[key] = value; };
```
on the minus side, it performs a global highland `Stream` update, which means other consumers will be affected, too.

if at all possible, we should make it so that upgrades / plugins are restricted either (1) to all streams originating from a bespoke version of `_`, or (2) to a single chain of methods (everything below a given `.use(...)` call in my example above). the reason is that in any system beyond a given complexity, some module might inject some `.foo()` functionality that name-collides with some unrelated `.foo()` defined by some other highland user in the same process. given how npm resolves dependencies, such collisions may affect users depending on the order that `npm install x` commands have been done (i believe that, but am not sure). therefore, global prototype monkeypatching is nothing more than a last resort.

**Comment by loveencounterflow** [comment]
#90

**Comment by loveencounterflow** [comment]
@caolan clearly the chaining / fluent interface is a problem here. i had a hunch that maybe ES6 proxies could be relevant here. and i indeed found http://molily.de/weblog/ecmascript5 and http://forum.de.selfhtml.org/archiv/2011/3/t203744/#m1378561; quote:

"Der Trick beim Chaining ist ja einfach nur, immer das gleiche Objekt zurück zu geben. D.h. man könnte einfach für alle Objekte eine Methode, bzw. einen Getter definieren, der einem ein Proxy-Objekt zurück gibt, dessen Methoden  neben dem Aufruf der Original-Methode dieses tun"

roughly, "since the whole idea in a fluent interface is to pass on one and the same object, one could just as well pass on a Proxy object with a getter method that does exactly that (pass on the same object)". if feasable (i don't know yet), then there would be one part, quite independent of highland.js, that cares for method retrieval from a number of sources done the fluid way. an instance of the 'organizer', if you will, is then pre-loaded with a number of core highland methods. users can then extend that object by inserting / unshifting more name / method pairs to the top (for overrides) or pushing name / method pairs to the bottom (for fallbacks) of the name lookup chain.

basically,

```coffee
_ = new_fluid_proxy()
_ = _.use highland_core
_ = _.use highland_bells
_ = _.use highland_whistles
module.exports = _;
-----------------------------------------------------------
_ = require 'highland.js'
_ = _.use require 'highland-foo'
# prepend so we override other uses of this name:
_ = _.use_first 'brobdignify', ( push, next ) -> ...

_ [ 1, 2, 3, 4, ]
  .frobulate 42
  .pipe output
```

i also have the hunch that my idea from above—to modify the instance within the stream's chain—would be too difficult to implement, too fragile, and not even convenient or important enough to care about.
